app/src/main/java/com/android/bookswap/model/map/Geolocation.kt

```diff
@@ -0,0 +1,74 @@
+package com.android.bookswap.model.map
+
+import android.Manifest
+import android.annotation.SuppressLint
+import android.app.Activity
+import android.content.pm.PackageManager
+import android.os.Looper
+import androidx.compose.runtime.mutableDoubleStateOf
+import androidx.core.app.ActivityCompat
+import com.google.android.gms.location.*
+
+class Geolocation(private val activity: Activity) {
+
+  private val REQUEST_LOCATION_PERMISSION = 1
+  private lateinit var fusedLocationClient: FusedLocationProviderClient
+  val latitude = mutableDoubleStateOf(Double.NaN) // Change to MutableState
+  val longitude = mutableDoubleStateOf(Double.NaN)
+
+  // Location request settings
+  private val locationRequest: LocationRequest =
+      LocationRequest.create().apply {
+        interval = 10000 // Update interval in milliseconds
+        fastestInterval = 5000 // Fastest update interval in milliseconds
+        priority = LocationRequest.PRIORITY_HIGH_ACCURACY
+      }
+
+  private val locationCallback =
+      object : LocationCallback() {
+        override fun onLocationResult(p0: LocationResult) {
+          p0.lastLocation.let { location ->
+            // Handle the updated location here
+            latitude.doubleValue = location.latitude
+            longitude.doubleValue = location.longitude
+            // You can save this location or notify other parts of your app
+          }
+        }
+      }
+
+  init {
+    fusedLocationClient = LocationServices.getFusedLocationProviderClient(activity)
+  }
+
+  // Request location permissions
+  private fun requestLocationPermissions() {
+    val permissions =
+        arrayOf(
+            Manifest.permission.ACCESS_FINE_LOCATION,
+            Manifest.permission.ACCESS_COARSE_LOCATION,
+            Manifest.permission.ACCESS_BACKGROUND_LOCATION)
```

Comment:
Potential issue : ACCESS_BACKGROUND_LOCATION needs API LEVEL 29 which is not the case for all devices.  I find a stackOverflow that can help you for this issue [here](https://stackoverflow.com/questions/73034972/why-does-the-permissions-dialog-not-appear-when-i-ask-for-background-location)